keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,578 @@
+    Template: &v1apply.PodTemplate{
+      Spec: &v1apply.PodSpec{
+        Containers: []v1.Containers{
+          {
+            Name: ptr.StringPtr("nginx"),
+            Image: ptr.StringPtr("nginx:latest"),
+          },
+        }
+      },
+    },
+  },
+}
+```
+
+For this approach, developers need to use a library like
+https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
+to inline primitive literals.
+
+#### Alternative 2: Generated "builders"
+
+Example usage:
+
+```go
+&appsv1apply.Deployment().
+  ObjectMeta(&metav1apply.ObjectMeta().
+    Name("nginx-deployment")).
+  Spec(&appsv1apply.DeploymentSpec().
+    Replicas(0).
+    Template(
+      &v1apply.PodTemplate().
+        Spec(&v1apply.SetPodSpec().
+          Containers(v1apply.ContainerList{
+            v1apply.Container().
+              Name("nginx").
+              Image("nginx:1.14.2")
+            v1apply.Container().
+              Name("sidecar").
+          })
+        )
+      )
+    )
+  )
+```
+
+<<[UNRESOLVED @jpbetz ]>> 
+Finalize how setter functions are named if we go with this alternative. Options we
+are considering are: (1) Just use the field name (2) prefix field name with "Set" (3)
+prefix field name with "With".
+<<[/UNRESOLVED]>>
+
+#### Comparison of alternatives
+
+See https://github.com/kubernetes/kubernetes/pull/95988 for a working implementation
+of alterative 1 and https://github.com/jpbetz/kubernetes/tree/apply-client-go-builders
+for a working implementation of alternative 2.
+
+Of the two leading alternatives--"builders" and "structs with pointers"--we implemented
+prototypes of both. They had roughly equivalent performance, and no differences
+in their capabilities. The choice between the two is primarily one of aesthetics/ergonomics.
+
+Some of the feedback we have heard from the community:
+
+- "structs with pointers" feels more go idiomatic and more closely aligned with
+  the go structs used for Kubernetes types both for builtin types and by
+  Kubebuilder.
+- It's nice how "builders" are clearly visually distinct from the main go types.
+- Having to use a utility library to wrap literal values as pointers for the
+  "structs with pointers" is not a big deal. I'm already familiar
+  with having to do this in go and once I learn use a utility library for it
+  I'm all set.
+- The "builders" are awkward to use in an IDE. I felt like I was fighting with
+  my IDE to get chain function calls and organize them hierarchally as expected
+  by this approach.
+
+TODO: We are providing the developer community with a fork of controller-tools
+that will allow them to generate apply configuration types that have both the
+alternatives. Our goal is to get feedback from developers after they try out the
+generated apply configurations and use that to inform our decision.
+
+While it is possible to generate types that have both the pointer fields exposed
+and the builder functions, we would prefer to provide clear guidance to the
```

Comment:
I don't know that I'm sold on avoiding the option for each, but I wouldn't let that stop this from proceeding.